modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/repo/es/ElasticSearchDetectorMappingService.java

```diff
@@ -0,0 +1,377 @@
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.ConstantScoreQueryBuilder;
+import org.elasticsearch.index.query.NestedQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.query.RangeQueryBuilder;
+import org.elasticsearch.percolator.PercolateQueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static com.expedia.adaptivealerting.modelservice.repo.es.DetectorMappingEntity.LAST_MOD_TIME_KEYWORD;
+
+@Component
+@Slf4j
+@SuppressWarnings({"PMD.ExcessiveImports", "PMD.AvoidThrowingRawExceptionTypes"})
+public class ElasticSearchDetectorMappingService implements DetectorMappingService {
+    private static ObjectMapper objectmapper    = new ObjectMapper();
+
+    @Autowired
+    private ElasticSearchConfig elasticSearchConfig;
+
+    @Autowired
+    private ElasticSearchClient elasticSearchClient;
+
+    private final Timer delayTimer;
+    private final Counter exceptionCount;
+
+    @Autowired
+    public ElasticSearchDetectorMappingService(MetricRegistry metricRegistry) {
+        this.delayTimer = metricRegistry.timer("es-lookup.time-delay");
+        this.exceptionCount = metricRegistry.counter("es-lookup.exception");
+    }
+
+    @Override
+    public MatchingDetectorsResponse findMatchingDetectorMappings(List<Map<String, String>> tagsList) {
+        try {
+            List<BytesReference> refList = new ArrayList<>();
+            for (Map<String, String> tags: tagsList) {
+                XContentBuilder xContent = XContentFactory.jsonBuilder();
+                xContent.map(tags);
+                refList.add(BytesReference.bytes(xContent));
+            }
+            PercolateQueryBuilder percolateQuery =
+                new PercolateQueryBuilder(DetectorMappingEntity.QUERY_KEYWORD, elasticSearchConfig.getDocType(),
+                    refList, XContentType.JSON);
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+            ConstantScoreQueryBuilder constantScoreQueryBuilder = new ConstantScoreQueryBuilder(percolateQuery);
+            searchSourceBuilder.query(constantScoreQueryBuilder);
+            searchSourceBuilder.timeout(new TimeValue(elasticSearchConfig.getConnectionTimeout()));
+            //FIXME setting default result set size to 500.
+            searchSourceBuilder.size(500);
```

Comment:
What use case does this support? I know of one--given a metric, find its detectors. But that use case doesn't require a list of maps--it requires only a single map. Also 500 would be much larger than what's required for that use case (in most cases there would be just one, though I could imagine three or four being used--but 500 seems too large).

Is there another use case? Might be useful to document some sample use cases in the Javadoc.

Reply:
Updated comment on the use case.